### DISCUSSION TITLE
Release Candidate v3.4.0rc

### DISCUSSION THREAD

**Discussion by dewenni** [dewenni/ESP_Buderus_KM271#74] [Announcements]
# Release Candidate v3.4.0rc

## what's new

This Versions is all about messages and brings some great new features 🎉

### Messaging option - Pushover

There is now the possibility to get messages via Pushover. 
Depending on the parameter "Filter", you can define what kind of messages you want to receive.
In the settings you can find all necessary parameters to setup the client.

Each application, service, or utility that sends notifications through Pushover's API needs to have its own API token which uniquely identifies all of the API requests that it makes.
API tokens are free and can be registered through [Pushover website](https://pushover.net/apps/build).

![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/c652f39a-2b6a-4ad5-9b35-9ef55b502288)

### WebUI-Logger

There is also a new log function with which you can record various messages depending on the filter and display them via the WebUI. This can be useful for your own debugging and also for the further development of the software.
You can also configure the logging function and the associated filter in the settings.

![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/9799081f-f531-4d6b-abce-3ca3bcab34b9)

## changelog

- rework messaging function. mqtt publish function was replaced with a generic messaging function for further options.
- add messaging option for "Pushover" #31
- add WebUI-Logger #70
- bugfix for updating OneWire sensor values if not enabled
- remove mqtt message "info = {"burner":0,"pump":100,"ww_temp":57,"boiler_temp":38}"
- improve behavior for incoming mqtt messages without payload (this could lead to problems in the past)
- improve compiler warnings in ArduinoIDE #72

## Assets

> [!IMPORTANT]   
> Please only use the OTA update if you have already flashed v3.2.5 or v3.3.1 to the Board via the serial cable to activate the larger partition scheme.
> If you want to update from a version <= 3.2.4 to this version, please flash it by cable.
> If you have already flashed the v3.2.5 or v3.3.1 by cable, then you can also use the OTA update.

**bin-files for OTA and Flash**
[bin-files-v3.4.0rc.zip](https://github.com/dewenni/ESP_Buderus_KM271/files/14233981/bin-files-v3.4.0rc.zip)

**Comment by dewenni** [comment]
For all those interested, I have a preview version with new functions for you here. I would like to use the discussion here to get feedback on the new functions before I make an official release.
Feel free to test it out and give me feedback if possible (in German or English).

---

Für alle Interessierten habe ich hier für euch eine Vorabversion mit neuen Funktionen. Ich möchte den Weg über die Diskussion hier nutzen um Feedback zu den neuen Funktionen zu bekommen, bevor ich ein offizielles Release mache.
Testet es gerne aus und gebt mir wenn möglich Feedback dazu (deutsch oder englisch).

**Comment by dewenni** [comment]
thanks to the Feedback from **Franck78**, I have added the possibility to change the sorting order of the log entries.
It is included in the last commit.

<img width="1092" alt="image" src="https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/14672445-b3b0-4195-88c1-7ba8dcf16af9">

**Comment by JieffeEl** [comment]
For what it's worth, the mqtt message "info ..." has its use (at least for me).
I use it through NodeRed to extract the temperature and burner state and export it to my home automation system, Domoticz. 
I know I can find the same information in /status/boiler_temp and /status/burner_control but the "info" MQTT telegram has the advantage of being repeated at regular interval. In Domoticz, if a temperature  sensor doesn't receive data after a certain delay (15 minutes in my case), it signals an error. 
That means the temperature sensor will show an error when the Buderus is off or in night mode (when the temperature changes very slowly). That just a cosmetic issue but enough to trigger my OCD :)

**Reply by dewenni** [reply]
ok, thanks for the feedback.
I thought that this was not really used.
It also has the problem that it is sometimes sent after a restart before the values it contains have been read by the Logamatic. As a result, it could happen that "zero values" were sent.
But if there is a need, I can leave it as it is and perhaps see if I can somehow prevent the "zero values" from being sent after the restart.

**Reply by JieffeEl** [reply]
Great! Many thanks!

**Comment by Franck78** [comment]
Ok,
About the logger and your 100 lines overload:

You should avoid a lot a string moves.

Forget the array of strings. 

In the circular buffer, put simply the final message :

`snprintf (logbuffer, logbuffersize, "Message: %s<br>", message)
`
and manage the circular aspect of it of course with a second snprintf when needed

Then in "server->on("/logger", HTTP_GET, [](AsyncWebServerRequest *request)"

send the buffer in two chunks.

```
    AsyncWebServerResponse *response = request->beginChunkedResponse("text/html", [](uint8_t *buffer, size_t maxLen, size_t index) {

          firstchunk:
          //constant part
           bytesToWrite = sprintf((char *)buffer, "<!DOCTYPE html><html class=\"HTML\"><head><style>body{background:#444857;font-family:monospace,monospace;color:#ffffff;line-height:150%%;}</style></head><body>");
          //then, somewhere in the middle of logbuffer to it's end
          len = something
          middle = somewhereinlogbuffer
          memcpy(buffer+bytesToWrite , middle,  len); //the oldest line is partially erased by the latest. Maybe do it clean by searching the next<br>
           bytesToWrite+=len;
           //string ending
           *logbuffer+len = \0;

          lastchunk:
          //it always starts at beginning of the logbuffer
          len = something
          memcpy(buffer ,logbuffer ,  len); //the oldest line is partially erased by the latest. Maybe do it clean by searching 
          bytesToWrite+=len;
          memcpy (buffer+bytesToWrite, "</body></html>\0");
          bytesToWrite += strlen( "</body></html>");

```

  I can do it if needed and agree on 'optimization'

**Reply by dewenni** [reply]
I don't know if I understand your optimization.

In the actual implementation I don't "move" the strings. The function to add a new message to the buffer is quite simple.
There is only one string-copy to add the new message into the buffer. If the buffer is full, the oldest message will be overwritten.
The logData.lastline is some kind of "write pointer" that marks the line for the next message.

```
void addLogBuffer(const char *message){  
  if (strlen(message)!=0){
    snprintf(logData.buffer[logData.lastLine], sizeof(logData.buffer[logData.lastLine]), "[%s]  %s", getDateTimeString(), message);
    logData.lastLine = (logData.lastLine + 1) % MAX_LOG_LINES; // update the lastLine index in a circular manner
  }
}
```

And on the other side, the web server function reads the buffer in the same way. depending on the sorting order it will read line by line from the buffer. Starting from logData.lastline in one or the other direction.

The "request->beginChunkedResponse" is specially for reading larger content.
I asks cyclically for new data until the return value is NULL. In my case I provide one message with every call.
This is probably less stressful than splitting the entire buffer into just 2 chucks.

Regarding general buffer and array of strings.
My first approach was also to have only a simple buffer and to queue the messages together there, separated by a "\n".
This would allow the size of the buffer to be used more efficiently, but the overall handling seemed less efficient to me.
If the buffer is full and a new message is added, then you have to search for and delete the area of the oldest message, which is relatively complex. If the new message is longer than the oldest message, you even have to delete two old messages and move the rest. Overall, the current solution seems simpler and therefore more robust to me. Now it is only one string-copy. In the other case, several operations are necessary. Reading from the buffer also seems to me to be somewhat more complex if you want to split the reading.
However, if you think you have a better approach, I am generally open to it. You are welcome to rebuild it for yourself and test whether it really works better. To be honest, I don't see that at the moment

**Comment by Franck78** [comment]
code ready, need cleaning but enough to see how much buffer lines are handled before the ESP is down.
Do you want it ?

**Reply by dewenni** [reply]
Sure, let's see 😉

**Reply by Franck78** [reply]
[ESP_Buderus_KM271-main340rc1.zip](https://github.com/dewenni/ESP_Buderus_KM271/files/14302543/ESP_Buderus_KM271-main340rc1.zip)

it is not super clean yet

In 'main.cpp', there really is only a call to clearbuffer. But if you could definitely integrate the '#ifdef' removing the DRD (such an annoying thing), I would be happy.

Try to see if you can get the little LED effect, really helpful  when the thing is silently rebooting ;)

**Reply by dewenni** [reply]
>  But if you could definitely integrate the '#ifdef' removing the DRD (such an annoying thing), I would be happy.

We need a possibility to set the software in "Setup-Mode" if something is wrong with the configuration.
Otherwise you have to connect it by cable in the case.
And I wanted a solution that works as hardware-independently as possible.
Nearly every ESP actually has a reset button. However, other buttons like the one on Daniel's board are optional.
That's why I found the solution with the DRD quite ok, even if it does have a few disadvantages.
You can't read out the reset button from the ESP, can you?
Otherwise I could also check whether the reset button has been pressed for longer than 3 seconds. But I don't think that works.

**Reply by Franck78** [reply]
oh, I'm not asking to remove anything. With the ifdef , it requires only a single 0 or 1 to remove DRD when needed. Let it included by default.

**Reply by dewenni** [reply]
So, I have implemented your changes and did some tests. I have also implemented a new `SIM_MODE` to get real data in a testing environment. I have a stand-alone ESP32 for testing before I go onto the Board that is installed in the logamatic.
The SIM_MODE can be activated in config.h with a simple `#define SIM_MODE true  `
You can start sending the simdata direct in the Logger Tab of the webUI or with a new mqtt command "../cmd/simdata"

**Regarding performance**
I found the same problems with your implementation as with my previous implementation.
With 5kb buffer it works quite stable and with 10kb buffer there are problems. Either the ESP crashes or the mqtt connection breaks down and the webUI is no longer accessible. If there is a crash, you can usually see references to the HEAP or AsycTCP / AsyncWebserver functions in the backtrace.

here are some examples with 10kb buffer in your implementation:

example 1:
```
Buffer too small not handled Maxlen: 5736 chunkSize:7932

assert failed: block_trim_free heap_tlsf.c:377 (block_is_free(block) && "block must be free")

  #0  0x40083bb9:0x3ffb3380 in panic_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/panic.c:402
  #1  0x4008d9c9:0x3ffb33a0 in esp_system_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/esp_system.c:128
  #2  0x400933c9:0x3ffb33c0 in __assert_func at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/assert.c:85
  #3  0x40092515:0x3ffb34f0 in block_locate_free at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/heap/heap_tlsf.c:447
      (inlined by) tlsf_malloc at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/heap/heap_tlsf.c:779
  #4  0x40092d5b:0x3ffb3510 in multi_heap_malloc_impl at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/heap/multi_heap.c:197
  #5  0x40092fe8:0x3ffb3530 in multi_heap_malloc at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/heap/multi_heap_poisoning.c:230
```

example 2:
```
#0  0x400eaaf6:0x3fff43c0 in AsyncWebSocketClient::_onData(void*, unsigned int) at .pio/libdeps/esp32/ESP Async WebServer/src/AsyncWebSocket.cpp:641
  #1  0x400ead19:0x3fff4400 in std::_Function_handler<void (void*, AsyncClient*, void*, unsigned int), AsyncWebSocketClient::AsyncWebSocketClient(AsyncWebServerRequest*, AsyncWebSocket*)::{lambda(void*, AsyncClient*, void*, unsigned int)#7}>::_M_invoke(std::_Any_data const&, void*&&, AsyncClient*&&, std::_Any_data const&, unsigned int&&) at .pio/libdeps/esp32/ESP Async WebServer/src/AsyncWebSocket.cpp:483
      (inlined by) _M_invoke at /Users/tonsjosts/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/xtensa-esp32-elf/include/c++/8.4.0/bits/std_function.h:297
  #2  0x40184c65:0x3fff4420 in std::function<void (void*, AsyncClient*, void*, unsigned int)>::operator()(void*, AsyncClient*, void*, unsigned int) const at /Users/tonsjosts/.platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/xtensa-esp32-elf/include/c++/8.4.0/bits/std_function.h:687
      (inlined by) AsyncClient::_recv(tcp_pcb*, pbuf*, signed char) at .pio/libdeps/esp32/AsyncTCP/src/AsyncTCP.cpp:915
  #3  0x40184cdd:0x3fff4450 in AsyncClient::_s_recv(void*, tcp_pcb*, pbuf*, signed char) at .pio/libdeps/esp32/AsyncTCP/src/AsyncTCP.cpp:1191
  #4  0x40185496:0x3fff4470 in _async_service_task(void*) at .pio/libdeps/esp32/AsyncTCP/src/AsyncTCP.cpp:159
      (inlined by) _async_service_task at .pio/libdeps/esp32/AsyncTCP/src/AsyncTCP.cpp:194
```

I think you can easily reproduce it when you start sending the Sim-Data by pressing the button in the webUI and after some seconds, press the refresh button.

Maybe the `beginChunkedResponse` is still not the best way for our use case. I'll experiment a bit more with the other methods.

attached the source files of the merged files and my extension with the simulation data
 
[3.4.0rc_franck78_SimMode.zip](https://github.com/dewenni/ESP_Buderus_KM271/files/14321903/3.4.0rc_franck78_SimMode.zip)

**Reply by Franck78** [reply]
It's normal for this version to segfault with 10K buffer. The 10Kb buffer cannot be split , "beginChunkedResponse" provides something around 5kb buffer. 

I rewrote it correctly (can handle 10K buffer no problem).
It was quick doing to see if sending 100 small chunk is better than sending 2 bigger (it should because calling the IP stack for only 100bytes is costly).

What I experience with >10K buffer is strange behavior of the Web server. Sometimes slow, Connect/Reconnect, but no segfault. 
I don't really know how to check this, find what is happening to the RTOS (the heart od ESP).
I will see it SIM_MOD helps me.



This code handles buffer and chunkSize as it should.

```
  server->on("/logger", HTTP_GET, [](AsyncWebServerRequest *request){
    AsyncWebServerResponse *response = request->beginChunkedResponse("text/html", [](uint8_t *buffer, size_t maxLen, size_t index) {
        static size_t half1, half2, intro, ending;
        size_t chunkSize;
        // read data from buffer to Chunk
        //Serial.printf("index (%u) full(%i)", index, logData.full);
        if (index == 0){ //start
            strcpy ((char *)buffer, "<!DOCTYPE html><html class=\"HTML\"><head><style>body{background:#444857;font-family:monospace,monospace;color:#ffffff;line-height:150%%;}</style></head><body>");
            intro = strlen ("<!DOCTYPE html><html class=\"HTML\"><head><style>body{background:#444857;font-family:monospace,monospace;color:#ffffff;line-height:150%%;}</style></head><body>");
            //when partial buffer, it contains nothing from tail to end of buffer
            if (logData.full)
                half2 = logData.buffer+BSIZE - logData.tail;
            else
                half2 = 0;
            half1 = logData.tail - logData.buffer;
            //fill the remaining buffer with half2 or half1
            chunkSize = min (half2, maxLen-intro);
            if (chunkSize) {
                memcpy (buffer+intro, logData.tail, chunkSize);
                //Serial.printf("BEGIN CH2 (%u)\n", chunkSize);
                half2 -= chunkSize;
            } else {
                chunkSize = min (half1, maxLen-intro);
                memcpy (buffer+intro, logData.buffer, chunkSize);
                //Serial.printf("BEGIN CH1 (%u)\n", chunkSize);
                half1 -= chunkSize;     
            }
            ending = strlen("</body></html>");
            return (intro+chunkSize);
        }

        // now send remainig half2 then half1 then ending html
        chunkSize = min (half2, maxLen);
        if (chunkSize) {
            memcpy (buffer, logData.tail, chunkSize);
            //Serial.printf("CH2 (%u)\n", chunkSize);
            half2 -= chunkSize;
            return chunkSize;
        }
        //half1
        chunkSize = min (half1, maxLen);
        memcpy (buffer, logData.buffer, chunkSize);
        //Serial.printf("CH1 (%u)\n", chunkSize);
        half1 -= chunkSize;
        //try to stuff in the htlm ending when half1 is done
        if (chunkSize+ending) {
            if (chunkSize+ending < maxLen) {
                //Serial.printf("sending ending (%u)\n", ending);
                strcpy ((char *)buffer+chunkSize, "</body></html>");
                chunkSize += ending;
                ending = 0;
            }
        }
        return chunkSize;
    });
    request->send(response);
  });

```



this one is less 'alpha'  looking ;)
```

/**
 * *******************************************************************
 * @brief   add new entry to LogBuffer
 * @param   none 
 * @return  none
 * *******************************************************************/
void addLogBuffer(const char *message){
  size_t messLen = strlen(message);

  if (messLen && (messLen<BSIZE )){
    //noInterrupts();
    //put the date
    //did DateString went past the end of the Buffer?
    size_t L = strlen ("[17.02.2024 - 12:45:55] ");  /* this is a constant*/
    *logData.tail  = '[';
    strcpy(logData.tail+1, getDateTimeString());
    strcat(logData.tail, "] ");
    //not enough room, copy char at beginning of buffer
    if (logData.tail + L > logData.buffer+BSIZE) {
        size_t Xtra = logData.tail + L -(logData.buffer+BSIZE);
        //Serial.printf ("Circ date  xtra=(%u) (%s)\n", Xtra, logData.buffer+BSIZE );
        logData.tail = logData.buffer;
        L = Xtra;
        memcpy (logData.buffer, logData.buffer+BSIZE,Xtra);
    }
    //interrupts();
    logData.tail += L;

    //now the message
    //messLen = strlen(message); 
    char* endMessage = logData.tail + messLen;
    if (endMessage > logData.buffer+BSIZE) {
       logData.full = true;
       size_t Xtra = endMessage-(logData. buffer+BSIZE);
       // Serial.printf ("Circ mess Buf return (%p) (%p) xtra=(%u)\n",logData.buffer+BSIZE , endMessage, Xtra);
       memcpy (logData.tail,  message, messLen-Xtra);
       memcpy (logData.buffer, message+messLen-Xtra, Xtra);
       logData.tail = logData.buffer+Xtra;
    } else {
      memcpy (logData.tail,  message, messLen);
      logData.tail += messLen;
    }

     //insert the newline <BR> 
    if ((logData.tail)>logData.buffer+BSIZE) {logData.full = true; logData.tail = logData.buffer;}
    *logData.tail++ = '<';
    if ((logData.tail)>logData.buffer+BSIZE) {logData.full = true; logData.tail = logData.buffer;}
     *logData.tail++ = 'B';
    if ((logData.tail)>logData.buffer+BSIZE) {logData.full = true; logData.tail = logData.buffer;}
    *logData.tail++ = 'R';
    if ((logData.tail)>logData.buffer+BSIZE) {logData.full = true; logData.tail = logData.buffer;}
     *logData.tail++ = '>';
    if ((logData.tail)>logData.buffer+BSIZE) {logData.full = true; logData.tail = logData.buffer;}

    //try to clear the last now broken line
    char* fill = logData.tail;
    while (*fill  != '<' &&  fill <= logData.buffer+BSIZE) {
      *fill++ = ' ';
    }
    //interrupts();
    
    //snprintf(logData.buffer[logData.lastLine], sizeof(logData.buffer[logData.lastLine]), "[%s]  %s", getDateTimeString(), message);
    //logData.lastLine = (logData.lastLine + 1) % MAX_LOG_LINES; // update the lastLine index in a circular manner
 
  }
}
```

**Reply by dewenni** [reply]
I have another remark to your solution. Could you move the `<br>`  from the buffer to the reading in webTools?
Otherwise the buffer is to specific for the use case of the webUI.
I am also playing around with Telnet and a use case could be to read the buffer with telnet.
In that case the `<br>` has to replaced with a simple line break `\n`.

**Reply by Franck78** [reply]
it would be ok if in the webpage the lines are inside a <  pre>< code> </ code></ pre> block. Can the WebUI do that ?
Changing \n to < BR> in the buffer is not possible (time consuming).

**Reply by Franck78** [reply]
ok, with pre/code inserted, it looks ok to me.

```

  server->on("/logger", HTTP_GET, [](AsyncWebServerRequest *request){
    AsyncWebServerResponse *response = request->beginChunkedResponse("text/html", [](uint8_t *buffer, size_t maxLen, size_t index) {
        static size_t half1, half2, intro, ending;
        size_t chunkSize;
        // read data from buffer to Chunk
        Serial.printf("index (%u) full(%i)", index, logData.full);
        if (index == 0){ //start
            strcpy ((char *)buffer, "<!DOCTYPE html><html class=\"HTML\"><head><style>body{background:#444857;font-family:monospace,monospace;color:#ffffff;line-height:150%%;}</style></head><body><pre><code>");
            intro = strlen ("<!DOCTYPE html><html class=\"HTML\"><head><style>body{background:#444857;font-family:monospace,monospace;color:#ffffff;line-height:150%%;}</style></head><body><pre><code>");
            //when partial buffer, it contains nothing from tail to end of buffer
            if (logData.full)
                half2 = logData.buffer+BSIZE - logData.tail;
            else
                half2 = 0;
            half1 = logData.tail - logData.buffer;
            //fill the remaining buffer with half2 or half1
            chunkSize = min (half2, maxLen-intro);
            if (chunkSize) {
                memcpy (buffer+intro, logData.tail, chunkSize);
                Serial.printf("BEGIN CH2 (%u)\n", chunkSize);
                half2 -= chunkSize;
            } else {
                chunkSize = min (half1, maxLen-intro);
                memcpy (buffer+intro, logData.buffer, chunkSize);
                Serial.printf("BEGIN CH1 (%u)\n", chunkSize);
                half1 -= chunkSize;     
            }
            ending = strlen("</code></pre></body></html>");
            return (intro+chunkSize);
        }

        // now send remainig half2 then half1 then ending html
        chunkSize = min (half2, maxLen);
        if (chunkSize) {
            memcpy (buffer, logData.tail, chunkSize);
            Serial.printf("CH2 (%u)\n", chunkSize);
            half2 -= chunkSize;
            return chunkSize;
        }
        //half1
        chunkSize = min (half1, maxLen);
        memcpy (buffer, logData.buffer, chunkSize);
        Serial.printf("CH1 (%u)\n", chunkSize);
        half1 -= chunkSize;
        //try to stuff in the htlm ending when half1 is done
        if (chunkSize+ending) {
            if (chunkSize+ending < maxLen) {
                Serial.printf("sending ending (%u)\n", ending);
                strcpy ((char *)buffer+chunkSize, "</code></pre></body></html>");
                chunkSize += ending;
                ending = 0;
            }
        }
        return chunkSize;
    });
    request->send(response);
  });

```

```
void addLogBuffer(const char *message){
  size_t messLen = strlen(message);

  if (messLen && (messLen<BSIZE )){
    //noInterrupts();
    //put the date
    //did DateString went past the end of the Buffer?
    size_t L = strlen ("[17.02.2024 - 12:45:55] ");  /* this is a constant*/
    *logData.tail  = '[';
    strcpy(logData.tail+1, getDateTimeString());
    strcat(logData.tail, "] ");
    //not enought room, copy char at beginning of buffer
    if (logData.tail + L > logData.buffer+BSIZE) {
        size_t Xtra = logData.tail + L -(logData.buffer+BSIZE);
        //Serial.printf ("Circ date  xtra=(%u) (%s)\n", Xtra, logData.buffer+BSIZE );
        logData.tail = logData.buffer;
        L = Xtra;
        memcpy (logData.buffer, logData.buffer+BSIZE,Xtra);
    }
    //interrupts();
    logData.tail += L;

    //now the message
    //messLen = strlen(message); 
    char* endMessage = logData.tail + messLen;
    if (endMessage > logData.buffer+BSIZE) {
       logData.full = true;
       size_t Xtra = endMessage-(logData. buffer+BSIZE);
       // Serial.printf ("Circ mess Buf return (%p) (%p) xtra=(%u)\n",logData.buffer+BSIZE , endMessage, Xtra);
       memcpy (logData.tail,  message, messLen-Xtra);
       memcpy (logData.buffer, message+messLen-Xtra, Xtra);
       logData.tail = logData.buffer+Xtra;
    } else {
      memcpy (logData.tail,  message, messLen);
      logData.tail += messLen;
    }

     //insert the newline 
    if ((logData.tail)>logData.buffer+BSIZE) {logData.full = true; logData.tail = logData.buffer;}
    *logData.tail++ = '\n';
    /*
    if ((logData.tail)>logData.buffer+BSIZE) {logData.full = true; logData.tail = logData.buffer;}
     *logData.tail++ = 'B';
    if ((logData.tail)>logData.buffer+BSIZE) {logData.full = true; logData.tail = logData.buffer;}
    *logData.tail++ = 'R';
    if ((logData.tail)>logData.buffer+BSIZE) {logData.full = true; logData.tail = logData.buffer;}
     *logData.tail++ = '>';
    if ((logData.tail)>logData.buffer+BSIZE) {logData.full = true; logData.tail = logData.buffer;}
    */

    //try to clear the last now broken line
    char* fill = logData.tail;
    while (*fill  != '\n' &&  fill <= logData.buffer+BSIZE) {
      *fill++ = ' ';
    }
    //interrupts();
    
    //snprintf(logData.buffer[logData.lastLine], sizeof(logData.buffer[logData.lastLine]), "[%s]  %s", getDateTimeString(), message);
    //logData.lastLine = (logData.lastLine + 1) % MAX_LOG_LINES; // update the lastLine index in a circular manner
 
  }
}

```

**Comment by solutionfound** [comment]
Today I gave your RC a try in the following sequence:
I came from 3.2.3 and flashed to 3.2.5 with the esp version of the software by debug interface - starting at 0x0. - I uploaded the old config.jason which I did not manage in the access point mode therefore I added the wifi data and did it after the card was accessible in the normal wifi network. 
Then I flashed OTA the 3.3.1 version and tested it in the Buderus which worked.
Then I flashed the 3.4.0 rc OTA - there I had to do a power cycle as it did not restart.
Not connected to the Buderus I can see the web interface.
Do I insert the KM217wifi into the Buderus 2017 the web interface is (again) not accessible with similar symptoms like the version 2.2.4 had.
To be extra sure I flashed the 3.4.0rc via the debug port with the esp version of the software. The symptoms were completely the same.
The MQTT part was working as I got messages to my OpenHab server. Only the web front end is not accessible.

After that I tried to downgrade over the web front end - not inserted into the Buderus - and just powered on the card using the debug header as power supply. I could upload the software with the web front end but the card did not do a reset. When power cycle it the 3.40rc firmware still was installed and booted.
Finally I flashed the 3.3.1 again via debug interface where the web front end shows no errors.

I was so keen on having the 3.4.0rc as I would then had a convenient way to investigate on the values for the solar part of the Buderus remotely with your new log feature page. And I wanted to contribute in testing the version.
Unfortunately it seems I need to wait a little until the version has stabilized as I am usually 100km away from the Buderus and as long as the OTA is not working completely reliable I'd better not flash versions remotely.

But already now it is a superb software & hardware and looking forward on the further progress. 

By the way for my gas Buderus with ems bus I use this software - perhaps interesting to see how the representation and features are implemented also using an esp32

https://emsesp.github.io/docs/

There is a live demo 
https://ems-esp.derbyshire.nl/
Account: admin - password: ems-esp
